### PR TITLE
test(functional-tests): audit subscription-tests functional tests

### PR DESCRIPTION
**Pull request by Trinaa**
## Because

the functional tests have a nightly 90-day success rate < 95% and a skip rate ~ 11-12% indicating waning stability and sustainability respectively.

## This pull request

audits the 'subscription-tests' functional test module making changes to improve these metrics:
1. remove test case 'support form with active subscriptions > go to support form, submits the form
    - The test case can present as a false positive since it can't be fulfilled due to Zendesk limitations in CI. Furthermore, the steps it does execute are repeated in 'support form with active subscriptions > go to support form, cancel, redirects to subscription management' so we can reduce CI cost and execution time by removing it.
2. uncomment and add fixme annotation to  'Flow, acquisition and new user checkout funnel metrics > New user checkout URL to have RP-provided flow params, acquisition params & verify funnel metrics'
3. uncomment 'resubscription test > update mode of payment for paypal' and reduce scope to respect the boundary between SubPlat and PayPal UI to reduce intermittency risk in CI.
4. isolate the MetricsObserver to tests where it is relevant
5. remove duplicate `severity-2` describe groups
6. fix lint errors and warnings

## Issue that this pull request solves

Closes: [FXA-8954](https://mozilla-hub.atlassian.net/browse/FXA-8954)

## Checklist

_Put an `x` in the boxes that apply_

- [x] My commit is GPG signed.
- [x] If applicable, I have modified or added tests which pass locally.
- [x] I have added necessary documentation (if appropriate).
- [x] I have verified that my changes render correctly in RTL (if appropriate).

## Screenshots (Optional)

Please attach the screenshots of the changes made in case of change in user interface.

## Other information (Optional)

Any other information that is important to this pull request.

- I suggest to 'Hide Widespace' as one of the commits required shifting the code left
- Changes with explanations are split by commit, so I suggest going through them one by one. I will squash when merging. 

[FXA-8954]: https://mozilla-hub.atlassian.net/browse/FXA-8954?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: packages/functional-tests/tests/subscription-tests/coupons-ui-tests/resubscription.spec.ts [hunk 1]
```diff
@@ -148,8 +145,7 @@ test.describe('severity-2 #smoke', () => {
       expect(await subscriptionManagement.getCardInfo()).toContain('4444');
     });
 
-    //Disable as paypal sandbox is taking a long time to load
-    /*test('update mode of payment for paypal', async ({
+    test('update mode of payment for paypal', async ({
```
Comment (Trinaa) [inline packages/functional-tests/tests/subscription-tests/coupons-ui-tests/resubscription.spec.ts:148]: This test has been disabled since 2023/05/10, I'm thinking due to instability when executing steps in the PayPal interface? Currently L189 has a `waitForTimeout` that is raising a lint warning.

What is our level of involvement with the PayPal interface? I'm inclined to remove this test because I'm under the impression we are testing a third party, which should be their responsibility. Changes they make could break our CI without warning and not necessarily mean a break in functionality (false negatives).

File: packages/functional-tests/tests/subscription-tests/support.spec.ts [hunk 2]
```diff
@@ -37,45 +37,17 @@ test.describe('severity-1 #smoke', () => {
...
-      const subscriptionPage = await settings.clickPaidSubscriptions();
-      subscriptionManagement.page = subscriptionPage;
-      await subscriptionManagement.fillSupportForm();
-
-      //Since we don't have proper Zendesk config in CircleCI, the form cannot be successfully submitted
-      //await subscriptionManagement.submitSupportForm();
-      //expect(await subscriptionManagement.subscriptionManagementHeader()).toBe(true);
-    });
+    // Since we don't have a proper Zendesk config in CircleCI, the test
+    // case where a form is successfully submitted cannot be covered.
```
Comment (Trinaa) [inline packages/functional-tests/tests/subscription-tests/support.spec.ts:41]: Is there more information on the Zendesk limitation?

File: packages/functional-tests/tests/subscription-tests/subscription.spec.ts [hunk 3]
```diff
@@ -94,147 +94,141 @@ test.describe('severity-2 #smoke', () => {
...
+    }, { project }) => {
+      test.skip(
+        project.name === 'production',
+        'test plan not yet available in prod'
+      );
+      // Go to settings page and verify the Data Collection toggle switch is visible
+      await settings.goto();
+      expect(await settings.dataCollection.isToggleSwitch()).toBe(true);
+
+      // Toggle swtich, verify the alert bar appears and its status
```
Comment (pdehaan) [inline packages/functional-tests/tests/subscription-tests/subscription.spec.ts:112]: Prior art, but the 👀 see what the 👀 see.
```suggestion
      // Toggle switch, verify the alert bar appears and its status
```

File: packages/functional-tests/tests/subscription-tests/subscription.spec.ts [hunk 4]
```diff
@@ -94,147 +94,141 @@ test.describe('severity-2 #smoke', () => {
...
-    let metricsObserver: MetricsObserver;
-
-    test.beforeEach(({ pages: { subscribe } }) => {
+    test.beforeEach(() => {
       test.slow();
-      metricsObserver = new MetricsObserver(subscribe);
-      metricsObserver.startTracking();
     });
 
-    test.describe('severity-2', () => {
```
Comment (pdehaan) [inline packages/functional-tests/tests/subscription-tests/subscription.spec.ts:105]: Not sure I 100% follow what's happening here. Are we losing the `severity-2` metadata here, and if so, what impact does that have? Fewer tests run in CI?

File: packages/functional-tests/tests/subscription-tests/coupons-ui-tests/ui-tests.spec.ts [hunk 5]
```diff
@@ -7,22 +7,18 @@ import { MetricsObserver } from '../../../lib/metrics';
...
-
     test('verify plan change funnel metrics & coupon feature not available when changing plans', async ({
-      page,
       pages: { relier, subscribe },
     }, { project }) => {
       test.skip(
         project.name === 'production',
         'no real payment method available in prod'
       );
+      test.slow();
```
Comment (pdehaan) [inline packages/functional-tests/tests/subscription-tests/coupons-ui-tests/ui-tests.spec.ts:17]: Did we just move this from the `beforeEach()` hook because it was the only test in the suite?

File: packages/functional-tests/tests/subscription-tests/subscription.spec.ts [hunk 6]
```diff
@@ -94,147 +94,141 @@ test.describe('severity-2 #smoke', () => {
...
-          project.name === 'production',
-          'test plan not yet available in prod'
-        );
-        await settings.goto();
-        await settings.signOut();
-        await relier.goto();
-        await relier.clickSubscribe6Month();
-        expect(page.url()).toContain('&flow_begin_time=');
-      });
+      // Toggle swtich and verify its status
```
Comment (pdehaan) [inline packages/functional-tests/tests/subscription-tests/subscription.spec.ts:134]:
```suggestion
      // Toggle switch and verify its status
```

File: packages/functional-tests/tests/subscription-tests/support.spec.ts [hunk 7]
```diff
@@ -37,45 +37,17 @@ test.describe('severity-1 #smoke', () => {
...
-      test.slow();
-    });
-
-    test('go to support form, submits the form', async ({
-      pages: { login, relier, subscribe, settings, subscriptionManagement },
-    }, { project }) => {
-      test.skip(
-        project.name === 'production',
-        'no real payment method available in prod'
-      );
```
Comment (pdehaan) [inline packages/functional-tests/tests/subscription-tests/support.spec.ts:50]: Why delete the test? Wouldnt this still be running in local/stage (based on conditional skip)?

File: packages/functional-tests/tests/subscription-tests/coupons-ui-tests/ui-tests.spec.ts (hunk 5 above)
Comment (Trinaa) [reply packages/functional-tests/tests/subscription-tests/coupons-ui-tests/ui-tests.spec.ts:17]: Yep that's right!

File: packages/functional-tests/tests/subscription-tests/subscription.spec.ts (hunk 4 above)
Comment (Trinaa) [reply packages/functional-tests/tests/subscription-tests/subscription.spec.ts:105]: If i understood correctly, this 'severity-2' was nested within another 'severity-2 #smoke', a duplication (see https://github.com/mozilla/fxa/pull/16303/commits/9d1c90ae9a0beddaedf57fb59d370f73cc5bbce0)

File: packages/functional-tests/tests/subscription-tests/support.spec.ts (hunk 7 above)
Comment (Trinaa) [reply packages/functional-tests/tests/subscription-tests/support.spec.ts:50]: I explain a bit in the commit description: https://github.com/mozilla/fxa/pull/16303/commits/f7bf5493221900f04e9cded85afbef672ba5db59

TLDR: imo it gives a false positive because the last lines aren't executed.

Comment (pdehaan) [reply packages/functional-tests/tests/subscription-tests/support.spec.ts:50]: Ah! Right, sorry, I mis-read the last line&hellip;
Although, maybe it does still verify _something_ by filling out a form that isn't submitted. Not the full submit+support pipeline, but at least that the support form appears and the fields are there as expected. Although _although_, now that I dig into it, it looks like the next test also has `await subscriptionManagement.fillSupportForm()` anyways. Whew, what a whirlwind of excitement!

https://github.com/mozilla/fxa/blob/1e5038201635ce43aec872ae94c6337aa89185f9/packages/functional-tests/pages/products/subscriptionManagement.ts#L78-L100

Comment (Trinaa) [reply packages/functional-tests/tests/subscription-tests/support.spec.ts:50]: Oops maybe i shouldn't TLDR and just copy my commit reasoning. I'm thinking you arrived at the same overall conclusion.

image

File: packages/functional-tests/tests/subscription-tests/support.spec.ts [hunk 8]
```diff
@@ -37,45 +37,17 @@ test.describe('severity-1 #smoke', () => {
...
-      await subscribe.submit();
-
-      //Login to FxA account
-      await login.goto();
-      await login.clickSignIn();
-      const subscriptionPage = await settings.clickPaidSubscriptions();
-      subscriptionManagement.page = subscriptionPage;
-      await subscriptionManagement.fillSupportForm();
-
-      //Since we don't have proper Zendesk config in CircleCI, the form cannot be successfully submitted
```
Comment (pdehaan) [inline packages/functional-tests/tests/subscription-tests/support.spec.ts:66]: Need to see if we can find the Zendesk ticket in Jira for who can help get this unstuck in staging.

File: packages/functional-tests/tests/subscription-tests/coupons-ui-tests/resubscription.spec.ts (hunk 1 above)
Comment (Trinaa) [reply packages/functional-tests/tests/subscription-tests/coupons-ui-tests/resubscription.spec.ts:148]: After discussions with team members, my understanding is that when changing payment methods, the boundary between the SubPlat and PayPal systems occurs when users press the 'Change' button.

image

By reducing the scope of the test to verifying that users are re-directed to PayPal when pressing 'Change' we maintain coverage while reducing intermittency risks in the CI.

File: packages/functional-tests/tests/subscription-tests/support.spec.ts (hunk 2 above)
Comment (Trinaa) [reply packages/functional-tests/tests/subscription-tests/support.spec.ts:41]: Asked and answered in the SubPlat Weekly meeting taking place on Monday Feb 5th 2024 (see [notes](https://docs.google.com/document/d/14U8gCJUOD8QaXDAvHY01e2quu0eDrVpajDBU5pAnQbs/edit)).

File: packages/functional-tests/tests/subscription-tests/support.spec.ts (hunk 8 above)
Comment (Trinaa) [reply packages/functional-tests/tests/subscription-tests/support.spec.ts:66]: I followed up on this in today's SubPlat weekly meeting (see [notes](https://docs.google.com/document/d/14U8gCJUOD8QaXDAvHY01e2quu0eDrVpajDBU5pAnQbs/edit)).  We were unable to find the ticket. Solving this issue is not currently on the roadmap and conflicts with plans to re-architect the support ticket (Zendesk) code. For now documenting the limitation in code is considered sufficient.

Comment (pdehaan) [reply packages/functional-tests/tests/subscription-tests/support.spec.ts:66]: Sure sure... So if the plan is that this is dead, +1 to deleting code.
If we think it will be back sooner rather than later do we want to just change this to using a more generic `test.skip()` so we're shamed by having _n_ skipped tests that we'll eventually need to rectify.

Also, if we're unable to find the Jira ticket (i tried looking during our morning 1:1:1 but failed finding anything obvs for Zendesk) I say we refile the ticket so we can track it as a work item and smash that like and subscribe.

Comment (sardesam) [reply packages/functional-tests/tests/subscription-tests/support.spec.ts:66]: I think the issue is that we have a lot of tests that haven't been revisited or documented in full as to why they are failing, possible resolutions, when and how to resolve, etc. By the time we do get to them, like this audit, we are unable to go back and find these necessary details or related tickets.

Since this is an issue that isn't on the roadmap and thus will not be resolved in the near future, removing the code altogether and documenting this decision is a good approach.

If we are confident in how we can resolve this with little effort, then I would opt for a `test.fixme` as Kat has done in other tests, but in this case, since there is no viable solution right now, I am aligned with removing this code.

File: packages/functional-tests/tests/subscription-tests/subscription.spec.ts [hunk 9]
```diff
@@ -98,142 +98,138 @@ test.describe('severity-2 #smoke', () => {
...
-        await settings.signOut();
-        await login.clearCache();
-        await relier.goto();
-        await relier.clickSubscribe6Month();
-        expect(page.url()).toContain('&flow_begin_time=');
-      });
+      // Subscribe to plan and verify URL does not include flow parameter
+      await relier.goto();
+      await relier.clickSubscribe6Month();
+      expect(page.url()).toContain('&flow_begin_time=');
```
[the remark this answers is not in the document]
Comment (Trinaa) [reply packages/functional-tests/tests/subscription-tests/subscription.spec.ts:141]: That's a nice catch! I will fix!

**Review by sardesam — APPROVED** [review]
LGTM! 🚀
